# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/VideoStreamTrack.cpp
# Commit: c54cb3c10992
# Full Hash: c54cb3c109922bc15fb6d0ca67e5f2a05980c1a5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 2
# Description:
#   Bug 1493613 - Move MediaStream control from DOMMediaStream to MediaStreamTrack. r=padenot
#   
#   This is inherently large, because modifying these bits of DOMMediaStream and
#   MediaStreamTrack affects all consumers and producers of all DOMMediaStreams and
#   MediaStreamTracks.
# ==============================================================================

diff -r 0b03dd9d20ac -r c54cb3c10992 dom/media/VideoStreamTrack.cpp
--- a/dom/media/VideoStreamTrack.cpp	Wed Jul 31 09:35:06 2019 +0000
+++ b/dom/media/VideoStreamTrack.cpp	Wed Jul 31 07:58:17 2019 +0000
@@ -14,12 +14,12 @@
 namespace mozilla {
 namespace dom {
 
-VideoStreamTrack::VideoStreamTrack(DOMMediaStream* aStream, TrackID aTrackID,
-                                   TrackID aInputTrackID,
+VideoStreamTrack::VideoStreamTrack(nsPIDOMWindowInner* aWindow,
+                                   MediaStream* aInputStream, TrackID aTrackID,
                                    MediaStreamTrackSource* aSource,
                                    const MediaTrackConstraints& aConstraints)
-    : MediaStreamTrack(aStream, aTrackID, aInputTrackID, aSource,
-                       aConstraints) {}
+    : MediaStreamTrack(aWindow, aInputStream, aTrackID, aSource, aConstraints) {
+}
 
 void VideoStreamTrack::Destroy() {
   mVideoOutputs.Clear();
@@ -27,6 +27,9 @@
 }
 
 void VideoStreamTrack::AddVideoOutput(VideoFrameContainer* aSink) {
+  if (Ended()) {
+    return;
+  }
   auto output = MakeRefPtr<VideoOutput>(
       aSink, nsGlobalWindowInner::Cast(GetParentObject())
                  ->AbstractMainThreadFor(TaskCategory::Other));
@@ -34,6 +37,9 @@
 }
 
 void VideoStreamTrack::AddVideoOutput(VideoOutput* aOutput) {
+  if (Ended()) {
+    return;
+  }
   for (const auto& output : mVideoOutputs) {
     if (output == aOutput) {
       MOZ_ASSERT_UNREACHABLE("A VideoOutput was already added");